craft_application/services/provider.py

```diff
@@ -116,6 +117,16 @@ def instance(
         base = self.get_base(base_name, instance_name=instance_name, **kwargs)
         provider = self.get_provider(name=self.__provider_name)
 
+        devel_bases = {
+            devel_base_info.devel_base for devel_base_info in DEVEL_BASE_INFOS
+        }
+        if base.alias in devel_bases:
+            emit.message(
+                "The development build-base should only be used for testing purposes, "
```

Comment:
I'm curious as to why this message is shown here, in the provider service, instead of in the project. If I build using `build-base: devel ` in destructive mode shouldn't I get the message too?

Reply:
Oh destructive-mode, good point.

It doesn't [belong](https://github.com/canonical/craft-application/pull/302#discussion_r1558138522) as a Project validator, where in the project do you think is should it go?  When the project is [loaded](https://github.com/canonical/craft-application/blob/607475231eb411098370918c7debaf91dbd9cb9f/craft_application/application.py#L244)?

That may not be the right place either.  I think the warning occur when executing the lifecycle and possibly remote-build.  Perhaps I could create a function in the Project to log a warning and call it in those services?

Reply:
On in that original comment I meant that I thought the warning was secondary to this PR/task (as in, it could have been a separate PR). I think it belongs in the Project, that's how Snapcraft and Rockcraft have it afaik

Reply:
Ah, I understand.  I added it because it was easy to upstream the `devel` warning alongside the error and let me entirely remove the validation functions in each application.  But you are right, it is secondary to this task.

I can drop the warning from this PR or move it back inside `def _validate_devel()` in craft-application.  Which would you prefer?

Reply:
move it back please :)